assets/javascripts/discourse/services/chat-message-visibility-observer.js

```diff
@@ -0,0 +1,35 @@
+import Service from "@ember/service";
+
+export default class ChatMessageVisibilityObserver extends Service {
+  constructor() {
+    super(...arguments);
+    this._setup();
+  }
+
+  _setup() {
+    const options = {
+      root: document,
+      rootMargin: "-10px",
+    };
+
+    const callback = (entries) => {
+      entries.forEach((entry) => {
+        entry.target.dataset.visible = entry.isIntersecting;
+      });
+    };
+
+    this.observer = new IntersectionObserver(callback, options);
+  }
+
+  willDestroy() {
+    this.observer.disconnect();
+  }
+
+  observe(element) {
+    this.observer.observe(element);
+  }
+
+  unobserve(element) {
+    this.observer.unobserve(element);
+  }
+}
```

Comment:
We should implement a `willDestroy()` method here which cleans up the `IntersectionObserver` instance. (services are created/destroyed at the beginning/end of every test)